pkg/volume/azure_dd/attacher.go

```
@@ -0,0 +1,250 @@
+			os.Remove(deviceMountPath)
+			return err
+		}
+	}
+	return nil
+}
+
+type azureDiskDetacher struct {
+	mounter       mount.Interface
+	azureProvider azureCloudProvider
+}
+
+var _ volume.Detacher = &azureDiskDetacher{}
+
+func (plugin *azureDataDiskPlugin) NewDetacher() (volume.Detacher, error) {
+	azure, err := getAzureCloudProvider(plugin.host.GetCloudProvider())
+	if err != nil {
+		return nil, err
+	}
+
+	return &azureDiskDetacher{
+		mounter:       plugin.host.GetMounter(),
+		azureProvider: azure,
+	}, nil
+}
+
+func (detacher *azureDiskDetacher) Detach(diskName string, hostName string) error {
+	if diskName == "" {
+		return fmt.Errorf("invalid disk to detach: %q", diskName)
+	}
+	instanceid, err := detacher.azureProvider.InstanceID(hostName)
+	if err != nil {
+		glog.V(2).Infof("no instance id for host %q, skip detaching", hostName)
+		return nil
+	}
+	if ind := strings.LastIndex(instanceid, "/"); ind >= 0 {
+		instanceid = instanceid[(ind + 1):]
+	}
+
+	glog.V(4).Infof("detach %v from host %q", diskName, instanceid)
+	err = detacher.azureProvider.DetachDiskByName(diskName, "" /* diskURI */, instanceid)
+	if err != nil {
+		glog.Errorf("failed to detach azure disk %q, err %v", diskName, err)
+	}
+
+	return err
+}
+
+func (detacher *azureDiskDetacher) WaitForDetach(devicePath string, timeout time.Duration) error {
+	return wait.Poll(checkSleepDuration, timeout, func() (bool, error) {
+		glog.V(4).Infof("Checking device %q is detached.", devicePath)
+		if pathExists, err := util.PathExists(devicePath); err != nil {
+			return false, fmt.Errorf("Error checking if device path exists: %v", err)
+		} else if !pathExists {
+			return true, nil
+		} else {
+			return false, nil
+		}
+	})
+}
+
+func (detacher *azureDiskDetacher) UnmountDevice(deviceMountPath string) error {
+	volume := path.Base(deviceMountPath)
+	if err := util.UnmountPath(deviceMountPath, detacher.mounter); err != nil {
+		glog.Errorf("Error unmounting %q: %v", volume, err)
+	}
+	return nil
```

Why are unmount errors swallowed?